File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        int n = nums.length;
        int s = 0;
        int e = n-1;
        int mid = 0;
        while(s<=e){
            mid = s+(e-s)/2;

            if(nums[mid]==target)return mid;
            else{
                //left sorted
                if(nums[s]<=nums[mid]){
                    if(target>=nums[s] && target<=nums[mid]){
                        e = mid - 1;
                    }
                    else{
                        s = mid + 1;
                    }
                }
                //right sorted
                else{
                    if(target>=nums[mid] && target<=nums[e]){
                        s = mid + 1;
                    }
                    else{
                        e = mid - 1;
                    }
                }
            }
        }
        return -1;
    }
}

File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        s = 0
        e = len(nums)-1

        while(s<e):
            mid = (int)(s+(e-s)/2)

            if nums[mid]==target:
                return mid
            
            if nums[mid]>=nums[0]:
                if target>=nums[0]:
                    if target<nums[mid]:
                        e = mid
                    else:
                        s = mid + 1
                else:
                    s = mid + 1
            else:
                if target>=nums[0]:
                    e = mid
                else:
                    if target<nums[mid]:
                        e = mid
                    else:
                        s = mid + 1
        
        if nums[e]==target:
            return e
        return -1